### PR TITLE
Initial provider implementation

### PR DESCRIPTION
**Pull request by joan41868**
**Description**:

This **PR**:
- Introduces `HederaDefaultProvider`- a class which contains `getBalance` and `getTransaction` methods
- Introduces `HederaNetworks` enum, which contains a preset of strings, acceptable by the `@hashgraph` sdk
- Disables some of the tests run in GH Actions

**Related issue(s)**:
#10 

Fixes #

#10 
Assists #11 with default implementations of `getBalance` and `getTransaction`, which will be moved to the `BaseProvider` while refactoring it in a future **PR**.

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/networks/src.ts/hedera_networks.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
+// @ts-ignore
+import {AccountID} from "@hashgraph/proto";
```
Comment (joan41868) [inline packages/networks/src.ts/hedera_networks.ts:2]: The same networks are encapsulated in the `@hashgraph/sdk` module, thus this file will be deleted in the next few commits.

File: playground.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+// main - run with ` ts-node playground.js `
+(async () => {
+    const provider = new HederaProvider("mainnet");
+    const solidityAddr = toSolidityAddress([0, 0, 98])
+    const bal = await provider.getBalance(solidityAddr);
+    console.log(bal); // outputs: BigNumber { _hex: '0x0131dcc4030d40', _isBigNumber: true }
+    console.log(bal.toNumber()); // outputs: 336299263359826
+})();
```
Comment (joan41868) [inline playground.ts:27]: @Daniel-K-Ivanov

File: packages/tests/src.ts/test-providers.ts [hunk 3]
```diff
@@ -1373,7 +1373,7 @@ describe("Resolve ENS avatar", function() {
         { title: "ipfs", name: "ipfs-avatar.tests.eth", value: "https:/\/gateway.ipfs.io/ipfs/QmQsQgpda6JAYkFoeVcj5iPbwV3xRcvaiXv3bhp1VuYUqw" },
```
Comment (joan41868) [inline packages/tests/src.ts/test-providers.ts:1373]: This file is from the previous github actions PR

File: .github/workflows/nodejs.yml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 name: Node.js CI
```
Comment (joan41868) [inline .github/workflows/nodejs.yml:1]: This file is from the previous GH Actions **PR**

File: packages/providers/src.ts/hedera-provider.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+                expected length 40, got length ${address.length}`);
+    }
+    const addr2: number[] = [];
+    addr.map(e => addr2.push(e.valueOf()));
+    const shard = Long.fromBytesBE([0, 0, 0, 0, ...addr2.slice(0, 4)]);
+    const realm = Long.fromBytesBE(Array.from(addr.slice(4, 12)));
+    const num = Long.fromBytesBE(Array.from(addr.slice(12, 20)));
+
+    return [shard, realm, num];
+}
```
Comment (Daniel-K-Ivanov) [inline packages/providers/src.ts/hedera-provider.ts:59]: This should be part of the utils package IMO (cc @alshapi )

File: package.json [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {
```
Comment (Daniel-K-Ivanov) [inline package.json:1]: I think that we should change the `common` section and the `repository.url` as-well

File: packages/providers/package.json [hunk 7]
```diff
@@ -68,5 +68,8 @@
   "sideEffects": false,
   "tarballHash": "0x2582ad9a43f5113ef9143cf4f5b35cc5fe0cc709861834d5f5236ab06b2f4476",
   "types": "./lib/index.d.ts",
-  "version": "5.5.0"
+  "version": "5.5.0",
+  "devDependencies": {
```
Comment (Daniel-K-Ivanov) [inline packages/providers/package.json:72]: I think that this should be part of the `dependencies` and not `devDependencies`

File: playground.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import {HederaProvider} from "@ethersproject/providers";
```
Comment (Daniel-K-Ivanov) [inline playground.ts:1]: Lets create `examples` folder and place different `ts` files in that example folder. We would have `example` file for `providers.ts` f.e

File: tsconfig.project.json [hunk 9]
```diff
@@ -6,7 +6,8 @@
       "dom"
     ],
     "preserveWatchOutput": true,
-    "pretty": false
+    "pretty": false,
+    "allowJs": true
```
Comment (Daniel-K-Ivanov) [inline tsconfig.project.json:10]: Why do we need to `allowJs`?

File: playground.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+    view.setUint32(0, shard);
+    view.setUint32(8, realm);
+    view.setUint32(16, num);
+
+    return encode(buffer);
+}
+
+// main - run with ` ts-node playground.js `
+(async () => {
```
Comment (Daniel-K-Ivanov) [inline playground.ts:21]: I think that we can have 2 examples:

1. Example with `0x0000000000...98` (it is not 98 but you got the point)
2. Example with `hethers.utils.getAddress("0.0.98")` (converts it to eth address) + `provider.getBalance(ethAddress)`

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
🎊

File: tsconfig.project.json (hunk 9 above)
Comment (joan41868) [reply tsconfig.project.json:10]: Because of the hedera **js** sdk

File: package.json (hunk 6 above)
Comment (joan41868) [reply package.json:1]: I'm not sure what is expected to be put in there, so I'll probably need some help with this.

File: examples/provider.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
+import { DefaultHederaProvider } from "@ethersproject/providers";
```
Comment (Daniel-K-Ivanov) [inline examples/provider.ts:1]: The implementation must be such that users do not import anything except for `ethers`, thus importing `hashgraph/sdk` is not something we want to end up with

File: package.json (hunk 6 above)
Comment (Daniel-K-Ivanov) [reply package.json:1]: We need to remove the `common` section only

File: packages/providers/src.ts/hedera-provider.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+ * As every provider, this one also gives us read-only access.
+ *
+ * Constructable with a string, which automatically resolves to a hedera network via the hashgraph SDK.
+ */
+export class DefaultHederaProvider extends BaseProvider {
+    private readonly hederaClient: Client;
+    private readonly hederaNetwork: string;
+
+    constructor(network: string) {
+        super('testnet');
```
Comment (Daniel-K-Ivanov) [inline packages/providers/src.ts/hedera-provider.ts:66]: I dont think that we should pass hardcoded `testnet` here

File: packages/providers/src.ts/hedera-provider.ts [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * The hedera provider uses the hashgraph module to establish a connection to the Hedera network.
+ * As every provider, this one also gives us read-only access.
+ *
+ * Constructable with a string, which automatically resolves to a hedera network via the hashgraph SDK.
+ */
+export class DefaultHederaProvider extends BaseProvider {
+    private readonly hederaClient: Client;
+    private readonly hederaNetwork: string;
```
Comment (Daniel-K-Ivanov) [inline packages/providers/src.ts/hedera-provider.ts:63]: The `baseProvider` has this info as parameter. I don't think that we need it here

File: packages/providers/src.ts/hedera-provider.ts [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Transaction record query implementation using the mirror node REST API.
+     *
+     * @param txId - id of the transaction to search for
+     */
+    async getTransaction(txId: string | Promise<string>): Promise<TransactionResponse> {
+        txId = await txId;
+        const ep = '/api/v1/transactions';
+        const url = resolveMirrorNetworkUrl(this.hederaNetwork);
+        // The following logic will be refactored/moved to tx.wait() method
```
Comment (Daniel-K-Ivanov) [inline packages/providers/src.ts/hedera-provider.ts:103]: Yes, lets directly return without polling

File: packages/providers/src.ts/hedera-provider.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+            // retry each 0.5 seconds
+            await sleep(500);
+            counter++;
+        }
+    }
+
+    /**
+     * Allows us to get the underlying gRPC client and execute gRPC calls.
+     */
+    public getClient(): Client {
```
Comment (Daniel-K-Ivanov) [inline packages/providers/src.ts/hedera-provider.ts:126]: We should not expose this. Everything should be wrapped into the `hedera-provider`

File: packages/tests/src.ts/test-providers.ts [hunk 16]
```diff
@@ -1393,3 +1396,19 @@ describe("Resolve ENS avatar", function() {
...
+      const accountConfig = { shard : BigInt(0), realm: BigInt(0), num: BigInt(98)};
+      const solAddr = getAddressFromAccount(accountConfig);
+      const balance = await provider.getBalance(solAddr);
+      // the balance of 0.0.98 cannot be negative
+      assert.strictEqual(true, balance.gte(0));
+   });
+
+   it("Gets the transaction record", async () => {
+
+   });
```
Comment (Daniel-K-Ivanov) [inline packages/tests/src.ts/test-providers.ts:1413]: nit: empty test

File: rollup-ancillary.config.js [hunk 17]
```diff
@@ -4,50 +4,51 @@ import resolve from 'rollup-plugin-node-resolve';
 import commonjs from 'rollup-plugin-commonjs';
 import _globals from 'rollup-plugin-node-globals';
 
-import { terser } from "rollup-plugin-terser";
+import {terser} from "rollup-plugin-terser";
 
 function getConfig(project, minify) {
```
Comment (Daniel-K-Ivanov) [inline rollup-ancillary.config.js:9]: Do we need to include these changes

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: packages/providers/src.ts/hedera-provider.ts (hunk 12 above)
Comment (Daniel-K-Ivanov) [reply packages/providers/src.ts/hedera-provider.ts:66]: We will resolve this in a separate PR

File: rollup-dist.config.js [hunk 18]
```diff
@@ -87,24 +87,26 @@ function getUmdConfig() {
         mainFields: [ "browser", "main" ]
     }));
     plugins.push(commonjs({ }));
+    // plugins.push(json())
```
Comment (Daniel-K-Ivanov) [inline rollup-dist.config.js:90]: remove this line

File: examples/provider.ts (hunk 11 above)
Comment (joan41868) [reply examples/provider.ts:1]: I think we should solve this in a separate PR, where we have refactored the `BaseProvider` interface

**Review by Daniel-K-Ivanov — APPROVED** [review]

**Pull request merged**